unix/vncserver/vncsession.c

```diff
@@ -511,14 +519,23 @@ main(int argc, char **argv)
 
     const char *username, *display;
 
-    if ((argc != 3) || (argv[2][0] != ':')) {
-        fprintf(stderr, "Syntax:\n");
-        fprintf(stderr, "    %s <username> <display>\n", argv[0]);
-        return EX_USAGE;
+    int opt, forking = 1;
+
+    while ((opt = getopt(argc, argv, "D")) != -1) {
+        switch (opt) {
+        case 'D':
+            forking = 0;
+            break;
+        default:
+            usage();
+        }
     }
 
-    username = argv[1];
-    display = argv[2];
+    if ((optind > 2) || (argc != optind + 2) || (argv[optind +1][0] != ':'))
```

Comment:
This looks good except for `(optind > 2)`. I don't really understand what that is checking for?

Reply:
The `optind` variable is used by `getopt(3)` as the value of the index in `argv[]` for the next option argument `getopt(3)` will process.  Since we have at most one option, the highest valid value for `optind` is 2. It is also valid for `optind` to be 1 if the argc is 3 and the other two non-option arguments are correct. So `optind > 2` is how we check that we don't have too many option arguments in `argv[]`.

That said, I think it is equivalent to use `argc > 4` in place of `optind > 2`, and if you consider that to be easier to understand, I can change that test for the same usage error that `optind > 2` is testing for with `arcg > 4` if you wish.

Reply:
I also just checked that this passes the usage tests I currently have: `vncsession -DD user :1` 

but this does not pass the usage tests: `vncsession -D -D user :1`.

So that `optind > 2` check is what makes `vncsession -D -D user :1` correctly fail the usage test.

I think the first one will still pass the usage test also if we use `argc > 4`, though, so changing it won't change anything, but if we remove the `optind > 2` test entirely, I think `vncsession -D -D user :1` would also pass the usage test. Maybe we need to add a test to make `vncsession -DD user :1` also fail the usage test.

I presume we want only `vncsession <user> :<display num>` and `vncsession -D <user> :<display num>` to be the only cases that pass the test.

I expect it would be safe, though, to remove the `optind > 2` check entirely and allow the `vncsession -D -D user :1` syntax to also pass the usage test.

Reply:
These are details of how options can be specified, which is what we have `getopt()` for. So let it take care of that and if it is happy, then we are happy. :)

So please remove that check. We already handle if `getopt()` gets upset about something.